Add functions to auto-detect and load any private key type

intgr opened the pull request:
Two new functions were created:
* load_private_key(key_str, password=None)
* load_private_key_file(filename, password=None)

Previously there was no way to auto-load any key type. Many downstream
users re-implemented this logic themselves, which is pointless
duplication and could easily become outdated when new key types are
added.

This functionality required a significant refactoring of key creation
code paths, but none of the public interfaces have changed.

intgr commented:
Copying initial feedback from **ploxiln**:

> Yes ... but my fork has new-openssh-format support, and that new format has the key type string sort of in the middle. I may want more refactoring in order to avoid copying a significant chunk of the new-format parsing code.
> 
> (There is also a bit of round-about-ness converting between a file-object and a list of lines, which I might prefer some refactoring of... for example, `load_private_key()` wraps the `key_str` into a `StringIO`, and then `load_private_key_from_file()` calls `.read()` to get the whole thing as a `str` in memory again for type detection, and then the first thing `_read_private_key()` does is `.readlines()` which gets another full read in memory but into a list of line strings ... seems like the the list-of-lines-in-memory should be the native/internal type passed around and used directly in most places. Or something like that.)

intgr commented:
Sorry, I commented on the wrong PR.

@intgr:
> It seems that for this feature to work without doing redundant work (doing multiple passes over the file content), it's necessary to split out significant parts of logic from the `PKey` class into stand-alone functions.
> 
> While reading the code, I didn't much like the ping-ponging of logic between `PKey` base class and its subclasses, so I think that's an improvement in general. Just making sure that you agree that's the way to go?

@ploxiln:
> The helper methods on the PKey class are very similar to just having helper functions in pkey.py ... but they also organize and reference constants like `_CIPHER_TABLE` on `self` ... so to reduce the amount of churn, I think you could continue to use pretty-much-static methods. I don't think a _big_ refactor is needed, just the bits at the beginnings of the functions, but the details are already different in the other fork so maybe the discussion should be there :)

intgr commented:
> so to reduce the amount of churn, I think you could continue to use pretty-much-static methods

I too would prefer to keep the current structure of code, but that would make it quite ugly.

Currently all of these methods are called already on the correct instance. The `*Key` instance exists before any parsing happens.

However, if we want `load_private_key` to work nicely, then the the *instance cannot exist* until very late in the loading process: for instance, new-format decryption must happen before we find out which `PKey` subclass to construct.

I could decorate the necessary methods as `@classmethod`, to call them before an instance is made, and pretend we're still OOP, but it seems more Pythonic to have functions.

Further, in order to construct the correct `*Key` type, I need to have an enumeration of all subclasses of `PKey`. It feels wrong to put that logic inside `PKey` itself -- it would make a circular dependency between `PKey` and each of its subclass -- surely an antipattern?

----

One alternative would be to not touch the current structure at all. `load_private_key` could just try all `PKey` types one by one and see which one doesn't throw an exception. But UGH! Then parsing and decryption would be done multiple times. :)

----

Your thoughts?

ploxiln commented:
Yeah, that's a good point, I'm OK with moving these helper functions out of the class, instead of adding `@classmethod`.

intgr commented on paramiko/rsakey.py:
```diff
@@ -26,26 +26,31 @@
+
     def __init__(self, msg=None, data=None, filename=None, password=None,
-                 key=None, file_obj=None):
+                 key=None, file_obj=None, raw=None):
```
All `*Key` constructors now take a `raw=` argument that's used internally to pass raw data from `load_private_key`.

Should I call this `_raw` to emphasize the fact that it's an internal API?

ploxiln replied:
Yeah, could rename to `_raw`.

But here's another idea: take a `file_data` instead, which is the full string. I think users would find it useful. I did say I didn't want to have the parsing code written multiple times, or read the file multiple times ... but I am not against running the parsing on the full string multiple times (once to get type, once to actually decrypt and load). But I guess users could just use the new `pkey.load_private_key()` anyway. So, just an idea.

ploxiln replied:
(in this idea I guess the new kwarg would be `key_str` instead of `file_data`, for consistency with `load_private_key()`)

intgr replied:
Personally I prefer not parsing the data twice. I think users should stop constructing the `*Key` object types directly and instead rely on `load_private_key*()` as well, so improving the old-style API seems like a non-goal.

ploxiln replied:
OK, let's call it `_raw` to show that it is internal API. After all, it's not really "raw" like bytes, it's half-processed ...

intgr replied:
Agreed, I was just making that change as I saw your reply about it :)

intgr commented on paramiko/rsakey.py:
```diff
@@ -26,26 +26,31 @@
     """
 
+    LEGACY_TYPE = 'RSA'
+    OPENSSH_TYPE_PREFIX = 'ssh-rsa'
```
The `@register_pkey_type` decorator reads these attributes and stores them in a global, which keeps track of all known key types.

What's slightly iffy is that if a key class is never imported, then `load_private_key` cannot know about it. :(

But this should not be a problem, since the main `paramiko/__init__.py` contains these imports.

intgr commented on paramiko/pkey.py:
```diff
@@ -52,36 +178,121 @@ def unpad(data):
     FORMAT_ORIGINAL = 1  # PKCS#1 for RSA or SEC1 for EC
     FORMAT_OPENSSH = 2  # newer "openssh-key-v1" format
-    BEGIN_TAG = re.compile(r'^-{5}BEGIN (RSA|DSA|EC|OPENSSH) PRIVATE KEY-{5}\s*$')
-    END_TAG = re.compile(r'^-{5}END (RSA|DSA|EC|OPENSSH) PRIVATE KEY-{5}\s*$')
```
`BEGIN_TAG` and `END_TAG` are the only non-underscore variables that I changed in a backwards-incompatible manner.

I really doubt anyone is relying on these in ther own code, what do you think?

ploxiln replied:
yeah, it is extremely unlikely that anyone would be referencing these
👍

intgr commented on paramiko/pkey.py:
```diff
@@ -52,36 +178,121 @@ def unpad(data):
+    return decryptor.update(data) + decryptor.finalize()
+
+
+def _read_private_key_new_format(data, password):
```
The `_read_private_key_old_format` and `_read_private_key_new_format` functions were extracted from the `PKey` class but otherwise changed very little.

intgr replied:
In fact, I have reverted the extraction of these two functions.

After looking at this with a fresh pair of eyes today, you were right, keeping these two functions in `PKey` isn't that bad and makes the diff **so much easier** to read, that by itself is a significant benefit.

(`_read_private_key` being the execption because I re-wrote that significantly into `_parse_openssh_pkey`)

intgr commented on paramiko/pkey.py:
```diff
@@ -39,6 +39,132 @@
+    """
+    Common parser for PEM-like OpenSSH private key formats.
+    """
+    match = OPENSSH_KEY_RE.search(key_str)
```
Sorry, I found the old parsing logic too ugly :)

I'm now using a regex for matching the entire key. No more splitting into lines and then joining back together logic.

ploxiln replied:
👍

intgr commented on paramiko/pkey.py:
```diff
@@ -39,6 +39,132 @@
+    return pkformat, typ, data
+
+
+def load_private_key_file(filename, password=None):
```
Mainly two new public APIs were added: `load_private_key` (string API) and `load_private_key_file`.

The `load_private_key_file_obj` was dropped because I don't think it's needed. A user can simply do the `.read()` themselves.

ploxiln replied:
👍

intgr commented:
This PR is now feature-complete, should be ready to start reviewing.

Will do some more checks hopefully tomorrow or Friday, then should be ready for merging..

ploxiln commented on paramiko/ecdsakey.py:
```diff
@@ -89,29 +89,35 @@ def get_by_key_length(self, key_length):
                 return curve
 
 
+@register_pkey_type
```
I think this is OK, but here's an alternate, more direct alternative: after defining the class, call this at the end of the source file:

```python
class ECDSAKey(PKey):
    ...

register_pkey_type(ECDSAKey, legacy_type="EC", openssh_type_prefix="ecdsa-sha2-")
```

just an idea

intgr replied:
I prefer the decorator myself. Something at the very end of a file is likely to be missed.

I could take the `legacy_type` and `openssh_type_prefix` as argument to the decorator, but since they're also accessed from within the class itself (in `_from_private_key`), I prefer them as explicit attributes.

ploxiln replied:
I see your point about keeping the class members `LEGACY_TYPE` etc, I didn't notice the other usage.

ploxiln commented on paramiko/pkey.py:
```diff
@@ -39,6 +39,107 @@
+
+# TODO: look at what documentation needs to be written
+# TODO: look for code in paramiko-ng internally that could use this
+# TODO: check code style with Black?
```
I'm not a fan of Black, and forked from upstream paramiko just before it was run across the codebase. I prefer just flake8/pycodestyle.

The main place inside paramiko that could benefit from these two new functions is `SSHClient._auth()`
https://github.com/ploxiln/paramiko-ng/blob/2.7.4/paramiko/client.py#L593-L598

intgr replied:
Right, I'll skip the Black part then. :)

Not trying to push it or anything, just FWIW: I started out not liking the Black style and auto-formatting in general. But after trialing it for the past year on a work project, it completely changed my mind. I feel like it frees up cognitive resources that were previously occupied by code formatting (in those projects I configure my IDE to auto-format on every save). It immediately shuts down any bikeshedding over space placement and whatnot. And new people on a project don't need to spend time on learning the project's particular code style, especially relevant for open source projects.

Adopting Black would also simplify porting patches between `paramiko` and `paramiko-ng`. But that doesn't seem to be a big issue in practice; resolving rebase conflicts probably took me 10 minutes on this original patch.

ploxiln replied:
I don't use an IDE. I make sure that my editor does not modify whitespace (or anything) on any lines which I did not intentionally edit.

The auto-formatter can't fix bad variable and function names, or bad code organization. And it gets some things obviously wrong. Trying to make the auto-formatter correct in 100% of cases is a much more challenging problem than just writing the code of interest with reasonable formatting manually. The churn to the codebase makes "git blame" much more annoying. I ported a bunch of PRs from upstream paramiko already (and most were pre-Black).

"go fmt" isn't so bad because it's a bit more like flake8 - it enforces indentation and other kinds of spacing, but lets you wrap lines in a few different ways. (But I still find it's choice of whitespace around nested arithmetic operators sometimes quite irritating.)

intgr replied:
> I make sure that my editor does not modify whitespace

That's explicitly not a use case of Black. Black is all-or-nothing: either everybody working on a project uses it, or nobody can reasonably use it.

> Trying to make the auto-formatter correct in 100% of cases

I assume by "correct" you mean "looks pleasing"? I agree sometimes there are cases where you have to help it along a bit, by extracting sub-expressions into variables. But that also tends to increase readability in general.

I think Black is mostly being held back by the obsolete line length limit of 88 chars (or worse, 79 in upstream paramiko). My preference in personal projects is 120 (with or without Black). `--line-length=120` gets rid of a lot of unnecessary line splitting with Black, at which point while sometimes worse than hand-tuned formatting, in some cases I find it's better than what I see humans produce, and ends up roughly on par. With the added benefit of absolute consistency, that's still a win.

> The churn to the codebase makes "git blame" much more annoying.

If everybody uses Black's code style then I find it actually reduces unnecessary churn (except for the initial conversion at first).

intgr commented:
Pushed a new significant update:
* I migrated a few more `pkey.py` private functions into classmethods in `PKey`. I'm still 50/50 on this, what do you think? I wouldn't mind moving them back. :)
* Renamed the `raw` argument to `_raw` in `PKey` subclasses.
* Wrote documentation for the new public functions. They now appear at the top of the "Key handling" documentation page.
  * Weirdly enough, the `RSAKey(filename=...)` etc constructor arguments were never documented.
* Converted most internal users (demos and tests) to use the new `load_private_key` function. With the exception of `test_pkey.py`, so old code paths are still excercised too.
  * TODO: Didn't convert `paramiko.client.SSHClient._key_from_filepath` yet.

ploxiln commented:
Converting demos I'm fine with, and maybe some parts of the tests that have nothing to do with keys, but a bunch of stuff in `test_client.py` and `test_transport.py` I would rather remained unchanged for now.

intgr commented:
Fair enough, reverted these two tests files.

ploxiln commented:
about ready to remove the TODO comments, and the wip tag? after that I'll just do some local manual sanity-testing before merge

intgr commented:
Yes :) Removed the TODOs, tweaked commit msg and a docstring.

The conversion for `paramiko.client.SSHClient._key_from_filepath` is non-trivial and it's not critical, it makes sense to do that as a separate PR.

intgr commented:
I also wanted to do a coverage run, to make sure all code I added has reasonable test coverage. But I can't get `pytest --cov-report=html` to work (also tried `--cov-report=html:$PWD/`, `--cov-report=html:cov_html`). Any ideas?

ploxiln commented:
> The conversion for paramiko.client.SSHClient._key_from_filepath is non-trivial and it's not critical

That's reasonable. But I think it's pretty easy (make `cls` optional), and doing that locally is how I plan to test this in real-world end-to-end situations. It's the only place in the actual parmiko codebase (as opposed to demos and tests) that can really use the new functionality. I'm happy to do it myself though, and I'll open a PR for you to take a look before merging it.

> But I can't get pytest --cov-report=html to work

Honestly, I usually do not bother with coverage, I'm not a fan. But I got it to work like so:

```
$ pip install -r dev-requirements.txt
$ pip install -u pytest coverage pytest-cov
$ pytest --cov-report=html --cov=paramiko tests/
...
---------- coverage: platform darwin, python 3.7.5-final-0 -----------
Coverage HTML written to dir htmlcov
```

ploxiln commented:
This is what I had in mind for SSHClient (un-tested):

```diff
--- a/paramiko/client.py
+++ b/paramiko/client.py
@@ -484,7 +484,7 @@ class SSHClient (ClosingContextManager):
         """
         return self._transport
 
-    def _key_from_filepath(self, filename, klass, password):
+    def _key_from_filepath(self, filename, klass=None, password=None):
         """
         Attempt to derive a `.PKey` from given string path ``filename``:
 
@@ -502,7 +502,10 @@ class SSHClient (ClosingContextManager):
             key_path = filename
             cert_path = filename + cert_suffix
         # Blindly try the key path; if no private key, nothing will work.
-        key = klass.from_private_key_file(key_path, password)
+        if klass:
+            key = klass.from_private_key_file(key_path, password)
+        else:
+            key = load_private_key_file(key_path, password)
         # TODO: change this to 'Loading' instead of 'Trying' sometime; probably
         # when #387 is released, since this is a critical log message users are
         # likely testing/filtering for (bah.)
@@ -591,19 +594,16 @@ class SSHClient (ClosingContextManager):
 
         if not two_factor and 'publickey' in allowed_types:
             for key_filename in key_filenames:
-                for pkey_class in (RSAKey, DSSKey, ECDSAKey, Ed25519Key):
-                    try:
-                        key = self._key_from_filepath(
-                            key_filename, pkey_class, passphrase,
-                        )
-                        allowed_types = set(
-                            self._transport.auth_publickey(username, key))
-                        two_factor = (allowed_types & two_factor_types)
-                        if not two_factor:
-                            return
-                        break
-                    except SSHException as e:
-                        saved_exception = e
+                try:
+                    key = self._key_from_filepath(key_filename, password=passphrase)
+
+                    allowed_types = set(self._transport.auth_publickey(username, key))
+                    two_factor = (allowed_types & two_factor_types)
+                    if not two_factor:
+                        return
+                    break
+                except SSHException as e:
+                    saved_exception = e
 
         if allow_agent and not two_factor and 'publickey' in allowed_types:
```

intgr commented:
Doh sorry, I was missing the `--cov=paramiko` option.

I'm happy with the coverage and didn't find any dead code.

Your provided diff for `_key_from_filepath` looks good to me.

intgr commented:
@ploxiln  Is the documentation at https://ploxiln.github.io/paramiko-ng/api/keys.html updated manually?

ploxiln commented:
Yes :) done now
